### PR TITLE
Enable more SpotBugs checks, ignore known issues

### PR DESCRIPTION
**Pull request by proski**
```
*  Enable more SpotBugs checks, ignore known issues
   
   Set SpotBugs threshold to Low. Use fb-contrib and find-sec-bugs plugins
   for SpotBugs.
   
   Add src/spotbugs/excludesFilter.xml file to ignore known issues. The
   parent POM makes SpotBugs use that file if it exists.
   
   It is better to check for more issues and ignore known issues than to
   risk adding new issues unknowingly.
```

I knew that I could set the threshold to Low or to enable plugins, but not both. Some of the Low issues detected by plugins are hard to fix. Then I realized that there is nothing wrong in having an exclusion file.

The plugin POM makes it easy. The exclusion file is found automatically if it's called `src/spotbugs/excludesFilter.xml`. We can remove that file one day, and we won't have to change `pom.xml`.

Many of the ignored issues are easy to fix, but that's not the point of this PR. It will be done separately. In fact, #137 will fix `SIC_INNER_SHOULD_BE_STATIC_ANON`, the only remaining non-plugin Low warning in the code.

In some cases, we can decide to ignore a specific rule (in particular, I don't think that all "mature" classes should implement `toString()`). Then we can add the rationale for the exclusion file as a comment.

The message about the missing `compare` class is bogus and appears to be similar to https://github.com/spotbugs/spotbugs/issues/527 (however, no lambdas are used in our `StashPullRequestComment#compare()` implementation). In any case, it doesn't affect the test result.

### CODE REVIEW

**Comment by proski** [comment]
I have already fixes for half of the issues. A few more a hard, but worth doing. One issue is actually a problem, a function returning null is marked as Nullable, and it's a recent change. I introduced that issue while simplifying a PR. That's why it's good to cut the noise down so we can see the new issues.

**Comment by proski** [comment]
Rebased and removed `AI_ANNOTATION_ISSUES_NEEDS_NULLABLE`, it's already fixed by #170.

File: src/spotbugs/excludesFilter.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<FindBugsFilter>
+
+  <Match>
```
Comment (jakub-bochenski) [inline src/spotbugs/excludesFilter.xml:10]: would be nice to get some short explanations for each exclusions, but I'm not insisting, especially if it's temporary

can you comment at least on the cases where there are open bugs in spotbugs or we don't intend to fix the violations in near future?

**Comment by jakub-bochenski** [comment]
> In some cases, we can decide to ignore a specific rule (in particular, I don't think that all "mature" classes should implement toString()). Then we can add the rationale for the exclusion file as a comment.

IMO `toString()` is just a debug convenience, you business logic shouldn't be relying on that. Based on that any problems with the method are low priority to me

**Comment by proski** [comment]
> > In some cases, we can decide to ignore a specific rule (in particular, I don't think that all "mature" classes should implement toString()). Then we can add the rationale for the exclusion file as a comment.
> 
> IMO `toString()` is just a debug convenience, you business logic shouldn't be relying on that. Based on that any problems with the method are low priority to me

I agree. That's one of the warnings that we should probably ignore.

File: src/spotbugs/excludesFilter.xml (hunk 1 above)
Comment (proski) [reply src/spotbugs/excludesFilter.xml:10]: I've added short explanations. I'll post a comment on the status of every exclusion.

**Comment by proski** [comment]
* CRLF_INJECTION_LOGS - we will fix is eventually by sanitizing inputs, but it's a lot of work.
* IMC_IMMATURE_CLASS_NO_TOSTRING - this will probably stay. Some classes could have toString(), but I don't think all classes should.
* IMC_IMMATURE_CLASS_WRONG_FIELD_ORDER - fix submitted as #172
* OPM_OVERLY_PERMISSIVE_METHOD - fix submitted as #173
* FCCD_FIND_CLASS_CIRCULAR_DEPENDENCY - probably fixable by splitting trigger and configuration. Could be tricky.
* PRMC_POSSIBLY_REDUNDANT_METHOD_CALLS - it's a false positive (the calls are to get current time), but "fixable" by moving the timestamps to a dedicated log function.
* HTTP_PARAMETER_POLLUTION - needs to be fixed by validating everything that goes into request URL. We should validate project and repository names in the configuration and at the runtime.
* PSC_PRESIZE_COLLECTIONS - probably a false positive. Can be suppressed easily, but we should still reexamine our use of data structures.
* WEM_WEAK_EXCEPTION_MESSAGING - not a big deal in presence of the cause. Still, we could add more data once we sanitize the URLs.

**Comment by jakub-bochenski** [comment]
Thanks, but I actually wanted this kind of comment to appear in the suppression file, instead of just explanations of the checks. I think they will be more visible there rather than in this PR

**Comment by proski** [comment]
> Thanks, but I actually wanted this kind of comment to appear in the suppression file, instead of just explanations of the checks. I think they will be more visible there rather than in this PR

OK, done.

**Comment by proski** [comment]
I started looking why `PSC_PRESIZE_COLLECTIONS` was issued for `getPullRequestComments()` but not for `getPullRequests()` and discovered that `getPullRequestComments()` is suboptimal. It uses two lists instead of one. One list collects half-processed data, the other list is made from it at the end. Once I changed that logic to process the comments as they arrive, the SpotBug warning went away. No pre-sizing of the array was actually needed. I'll post a PR soon, I just want to add more code coverage.

**Comment by proski** [comment]
Actually, the code coverage is good. Submitted #174.

**Pull request merged**